### ISSUE TITLE
oauth2 authorization code can be exchanged for access token multiple times

### ISSUE THREAD

**Issue by bradenrayhorn** [kanidm/kanidm#2775]
### I did this

It seems that a single authorization code can be exchanged multiple times for access tokens.

1. Setup an oauth2 application
2. Run OpenID Connect flow, upon receiving the callback containing the authorization code, exchange the code for an access token. (example below)

```
curl --location 'https://idm.example.com/oauth2/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'grant_type=authorization_code' \
--data-urlencode 'code=AUTHCODEHERE' \
--data-urlencode 'client_id=CLIENTIDHERE' \
--data-urlencode 'client_secret=CLIENTSECRETHERE' \
--data-urlencode 'redirect_uri=http://localhost:3000/auth/complete' \
--data-urlencode 'code_verifier=PKCEVERIFIERHERE'
```

The exact same request can be used multiple times to get multiple unique access tokens.

This seems to go against requirements from [RFC-6749 Section 4.1.2](https://www.rfc-editor.org/rfc/rfc6749#section-4.1.2):

> 
>    4.1.2 Authorization Response
> 
> ...
> 
>    code
>          REQUIRED.  The authorization code generated by the
>          authorization server.  The authorization code MUST expire
>          shortly after it is issued to mitigate the risk of leaks.  A
>          maximum authorization code lifetime of 10 minutes is
>          RECOMMENDED.  The client MUST NOT use the authorization code
>          more than once.  If an authorization code is used more than
>          once, the authorization server MUST deny the request and SHOULD
>          revoke (when possible) all tokens previously issued based on
>          that authorization code.  The authorization code is bound to
>          the client identifier and redirection URI.

Kanidm does not seem to deny subsequent requests using the same authorization code.

### I expected the following

The second request to exchange the code for an access token to fail.

### Kanidm version details

- Output of `kanidm(d) version`: `kanidmd 1.2.0`
- Are you running it in a container? If so, which image/tag?: Yes, `docker.io/kanidm/server:1.2.0`
- If not a container, how'd you install it: N/A
- Operating System / Version (On Unix please post the output of `uname -a`): N/A

### Any other comments

I might be misunderstanding the spec or be doing something else wrong, please let me know if that is the case.

**Comment by Firstyear** [comment]
Each authorisation code is only usable for 60 seconds. There is a fundamental challenge here that we need to deal with: namely, the fact we have replication.

Normally it's easy to solve some of these state related issues with replication with stick sessions on a load balancer, but that's not possible here because you have both the user-agent and the oauth2 client involved. 

So imagine your user-agent goes to the website (oauth2 client) and is redirected to kanidm. Via a load balancer you are sent to node A. This gives you the authorisation code which is then give back to the oauth2 client via a redirect.

At this point the oauth2 client contacts kanidm directly, and the load balancer is free to direct to any node in the topology. Here imagine it was sent to node B.

In this scenario it's plausible that these redirects and exchanges happen so fast that replication simply may not have occured yet. If we relied on something in the DB to replicate, it would make the authorisation code fragile and prone to erroring. 

At the same time, node B might handle the authorisation code and store in the DB that it was used, but node A doesn't know anything about this exchange (and may not for a number of minutes). Again, we have an authorisation code that can be reused. 

So instead we have to assume authorisation code's are stateless. They have short lifespan (60 seconds). We chose this to account for high latency internet links and other potential issues. It's a balance between fast expiry and allowing some room of error or delay. 

This is why you are seeing the behaviour - it's an intentional and necessary element in a distributed system. 

Now lets also consider mitigations here. Authorisation codes require one of two things:

* You must be a secure client with the client id and password
* You are a public client, and must have PKCE

In both cases, they provide ways to link the client who sent the authorisation code to the initial redirection. This means that were the authorisation code to leak, without the PKCE verifier *or* the client id + secret, the authorisation code can't be used anyway. 

Finally, each time an authorisation code is used, a unique oauth2 session record is made in the database, and linked to the originating session. This means that if the code were used multiple times, it's recorded and obvious that this happened. It also means that if the users browser session is logged out, all related oauth2 child sessions are also immediately invalidated regardless of how many there are.

So there was a lot of consideration put into these behaviours to come to the implementation we have. :)
